ui_test/src/tests.rs

```diff
@@ -29,25 +32,277 @@ fn main() {
     let comments = Comments::parse(&path, s);
     let mut errors = vec![];
     let config = config();
-    // Crucially, the intended error string *does* appear in this output, as a quote of the comment itself.
-    let stderr = br"
-error: Undefined Behavior: type validation failed: encountered a dangling reference (address 0x10 is unallocated)
-  --> tests/compile-fail/validity/dangling_ref1.rs:6:29
-   |
-LL |     let _x: &i32 = unsafe { mem::transmute(16usize) }; //~ ERROR encountered a dangling reference (address $HEX is unallocated)
-   |                             ^^^^^^^^^^^^^^^^^^^^^^^ type validation failed: encountered a dangling reference (address 0x10 is unallocated)
-   |
-   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
-   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
-           
-   = note: inside `main` at tests/compile-fail/validity/dangling_ref1.rs:6:29
-note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
-error: aborting due to previous error
-    ";
-    check_test_result(&path, &config, "", "", &comments, &mut errors, /*stdout*/ br"", stderr);
-    // The "OutputDiffers" is because we cannot open the .rs file
+    let stderr = r"this isn't checked anymore anyway and only used for test failure messages";
+    let diagnostics = Diagnostics {
+        rendered: filter_annotations_from_rendered(stderr).to_string(),
+        messages_from_unknown_file_or_line: vec![],
+        messages: vec![vec![], vec![], vec![], vec![], vec![], vec![
+            Message {
+                message:"Undefined Behavior: type validation failed: encountered a dangling reference (address 0x10 is unallocated)".to_string(),
+                level: Level::Error,
+            }
+        ]],
+    };
+    check_annotations(diagnostics, Path::new("moobar"), &mut errors, &config, "", &comments);
```

Comment:
Hm, this test doesn't really test much any more (in particular not what is used to test), does it?

Should `check_annotations` take the JSON so that we can make this a proper end-to-end test at least?

Reply:
Isn't the JSON easier to handle than the datatypes you are building by hand here currently?^^

But anyway, I am fine with keeping `check_annotations` like it is, but then this test specifically should pass a `diagnostics` that matches exactly what the file would produce (including the "renderd" part).

Reply:
> Isn't the JSON easier to handle than the datatypes you are building by hand here currently?^^

I guess I could use the `json!` macro to make writing them more convenient. But in either case, I want to check the json independently.



> but then this test specifically should pass a `diagnostics` that matches exactly what the previously problematic test case would produce (including the "renderd" part).

I thought about this, but it's fairly useless to test. the `rendered` part is unused for any decision making. We can replace it with an opaque type that implements just `Display` if we wanted to.

Reply:
The "rendered" part is compared with the reference file, right?

Reply:
In any case, the purpose of this test is exactly to ensure that the rendered part is not used for pattern matching decision making. :)

Reply:
Challenge accepted. `check_annotations` now does not take stderr as input anymore. It was only used for the case where an annotation did not find a corresponding diagnostic. It is now emitted seperately.